### PR TITLE
stop logging killing connection/stream because serving request timed out and response had been started

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: it stops persisting the following error in the server's log:
```
E0921 07:09:22.814963       1 runtime.go:78] Observed a panic: &errors.errorString{s:"killing connection/stream because serving request timed out and response had been started"} (killing connection/stream because serving request timed out and response had been started)
goroutine 1088 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2039b80, 0xc000552230)
	k8s.io/apimachinery@v0.19.0-rc.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc00257dc98, 0x1, 0x1)
	k8s.io/apimachinery@v0.19.0-rc.2/pkg/util/runtime/runtime.go:48 +0x89
panic(0x2039b80, 0xc000552230)
	runtime/panic.go:969 +0x175
k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).timeout(0xc0026077c0, 0xc002444460)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/timeout.go:257 +0x1bc
k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc0002df9a0, 0x26b7640, 0xc0005a7c00, 0xc00243ff00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/timeout.go:141 +0x2f3
k8s.io/apiserver/pkg/server/filters.WithWaitGroup.func1(0x26b7640, 0xc0005a7c00, 0xc00243fe00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/waitgroup.go:59 +0x137
net/http.HandlerFunc.ServeHTTP(0xc00127d650, 0x26b7640, 0xc0005a7c00, 0xc00243fe00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1(0x26b7640, 0xc0005a7c00, 0xc00243fd00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/requestinfo.go:39 +0x269
net/http.HandlerFunc.ServeHTTP(0xc00127d680, 0x26b7640, 0xc0005a7c00, 0xc00243fd00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithAuditAnnotations.func1(0x26b7640, 0xc0005a7c00, 0xc00243fc00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/audit_annotations.go:37 +0x142
net/http.HandlerFunc.ServeHTTP(0xc0002df9c0, 0x26b7640, 0xc0005a7c00, 0xc00243fc00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithWarningRecorder.func1(0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/warning.go:35 +0x1a7
net/http.HandlerFunc.ServeHTTP(0xc0002df9e0, 0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/endpoints/filters.WithCacheControl.func1(0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/endpoints/filters/cachecontrol.go:31 +0xa8
net/http.HandlerFunc.ServeHTTP(0xc0002dfa00, 0x26b7640, 0xc0005a7c00, 0xc00243fb00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server/httplog.WithLogging.func1(0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/httplog/httplog.go:91 +0x2f2
net/http.HandlerFunc.ServeHTTP(0xc0002dfa20, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1(0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/filters/wrap.go:51 +0xe6
net/http.HandlerFunc.ServeHTTP(0xc0002dfa40, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2042 +0x44
k8s.io/apiserver/pkg/server.(*APIServerHandler).ServeHTTP(0xc00127d710, 0x26aa080, 0xc002394110, 0xc00243fa00)
	k8s.io/apiserver@v0.19.0-rc.2/pkg/server/handler.go:189 +0x51
net/http.serverHandler.ServeHTTP(0xc002255c00, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:2843 +0xa3
net/http.initALPNRequest.ServeHTTP(0x26bd780, 0xc001dc88d0, 0xc00163b500, 0xc002255c00, 0x26aa080, 0xc002394110, 0xc00243fa00)
	net/http/server.go:3415 +0x8d
golang.org/x/net/http2.(*serverConn).runHandler(0xc001cd1080, 0xc002394110, 0xc00243fa00, 0xc0026076e0)
	golang.org/x/net@v0.0.0-20200602114024-627f9648deb9/http2/server.go:2147 +0x8b
created by golang.org/x/net/http2.(*serverConn).processHeaders
	golang.org/x/net@v0.0.0-20200602114024-627f9648deb9/http2/server.go:1881 +0x505
E0921 07:09:22.815015       1 wrap.go:39] apiserver panic'd on GET /openapi/v2
```

From now on the timeout handler will panic with `http.ErrAbortHandler` when a response to the client has been already sent and the timeout elapsed. As a result, the connection will be forcefully closed `EOF` and we will record `requestAbortsTotal`.

Previously the stack trace was persisted in the logs which caused confusion and false alarms.

Additionally, a new metric `requestAbortsTotal` was defined to count aborted requests. The new metric allows for aggregation for each `group`, `version`, `verb`, `resource`, `subresource` and `scope`.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**: Initially I wanted to use `errors.Is and errors.As` but that would require to have an additional function anyway as we need to cast `err interface{}` first. Other than that that would require coupling  `errConnKilled` with `http.ErrAbortHandler` which I didn't like.

**Does this PR introduce a user-facing change?**:

```release-note
A new metric `requestAbortsTotal` has been introduced that counts aborted requests for each `group`, `version`, `verb`, `resource`, `subresource` and `scope`.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/net/http_test.go [hunk 1]
```diff
@@ -401,8 +401,8 @@ func TestConnectWithRedirects(t *testing.T) {
 			}
 
 			netdialer := &net.Dialer{
-				Timeout:   wait.ForeverTestTimeout,
-				KeepAlive: wait.ForeverTestTimeout,
+				Timeout:   time.Second * 30,
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http_test.go:404]: why this? Seems to be unrelated.

File: staging/src/k8s.io/apimachinery/pkg/util/net/interface.go [hunk 2]
```diff
@@ -455,3 +456,23 @@ func ChooseBindAddressForInterface(intfName string) (net.IP, error) {
 	}
 	return nil, fmt.Errorf("unable to select an IP from %s network interface", intfName)
 }
+
+// AbortHandlerErrors holds a list of errors that signal the API server to abort and HTTP handler.
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/util/net/interface.go:460]: English parse error

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go [hunk 3]
```diff
@@ -28,13 +28,21 @@ import (
...
 	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 )
 
+// errConnKilled is thrown when a response has been already sent to the client and the handler timed out
+// note that panicking with errConnKilled also suppresses logging of a stack trace to the server's error log.
+// see also knet.AbortHandlerErrors
 var errConnKilled = fmt.Errorf("killing connection/stream because serving request timed out and response had been started")
 
+func init() {
+	knet.AbortHandlerErrors = append(knet.AbortHandlerErrors, errConnKilled)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:43]: this is ugly. Writing some global state....

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 4]
```diff
@@ -21,14 +21,15 @@ import (
...
+	"k8s.io/apimachinery/pkg/util/net"
 	"k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apiserver/pkg/server/httplog"
 )
 
 // WithPanicRecovery wraps an http Handler to recover and log panics (except in the special case of http.ErrAbortHandler panics, which suppress logging).
 func WithPanicRecovery(handler http.Handler) http.Handler {
 	return withPanicRecovery(handler, func(w http.ResponseWriter, req *http.Request, err interface{}) {
-		if err == http.ErrAbortHandler {
+		if net.IsAbortHandlerError(err) {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:32]: maybe we should at least log why we dropped the request?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:32]:
```
if errors.Is(http.ErrAbortHandler) {
  err := recover()
  klog.Error(err)
  panic(http.ErrAbortHandler)
}
```

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:32]: from golang docs `To abort a handler so the client sees an interrupted response but the server doesn't log an error, panic with the value ErrAbortHandler`

maybe we should actually throw `ErrAbortHandler` instead of `errConnKilled` ?
that would be seen as `  timeout_test.go:220: Get "http://127.0.0.1:61805": EOF` on the client side.

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 5]
```diff
@@ -29,11 +29,19 @@ import (
...
+			// If ServeHTTP panics, the server (the caller of ServeHTTP) assumes
+			// that the effect of the panic was isolated to the active request.
+			// It recovers the panic, logs a stack trace to the server error log,
+			// and either closes the network connection or sends an HTTP/2
+			// RST_STREAM, depending on the HTTP protocol. To abort a handler so
+			// the client sees an interrupted response but the server doesn't log
+			// an error, panic with the value ErrAbortHandler.
+			//
+			//
+			panic(http.ErrAbortHandler)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:44]: panicking with `http.ErrAbortHandler` (well any panic would do) will actually close the underlying connection.

Previously `errConnKilled` was captured, logged and the connection was left intact since calling `WriteHeader` multiple times is not supported by the std library.

I think that before closing the connection (before panicking) we could actually try to send a message to the client.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:44]: > I think that before closing the connection (before panicking) we could actually try to send a message to the client.

Even if we did, the msg would be ignored by the client (golang) library `On error, any Response can be ignored`
https://golang.org/src/net/http/client.go?s=20422:20474#L575

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:44]: Where did we catch errConnKilled ?

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 6]
```diff
@@ -29,11 +29,19 @@ import (
...
+			// If ServeHTTP panics, the server (the caller of ServeHTTP) assumes
+			// that the effect of the panic was isolated to the active request.
+			// It recovers the panic, logs a stack trace to the server error log,
+			// and either closes the network connection or sends an HTTP/2
+			// RST_STREAM, depending on the HTTP protocol. To abort a handler so
+			// the client sees an interrupted response but the server doesn't log
+			// an error, panic with the value ErrAbortHandler.
+			//
+			// We are throwing http.ErrAbortHandler deliberately so that a client is notified
+			klog.Warningf("Aborting the request since ErrAbortHandler was received (%v) on %v %v. This might result in a EOF error on the client side", err, req.Method, req.RequestURI)
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:43]: I'd prefer a metric, or a rate limit on this log message like that in https://github.com/kubernetes/kubernetes/pull/88600/files

**Comment by lavalamp** [comment]
/approve

I think my preference order is: metric, rate-limited log message, no log message, log message

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go (hunk 6 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:43]: +1 for metric.

What do we have now for errors like EOF or tls errors with `--v=2`? They are also visible, aren't they? A warning here is definitely too heavy. We won't warn about EOFs either, will we?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:43]: What will we see in `--v=2` for timeouts in general?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 7]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
...
+	// make another req to the server
+	// the connection should be reused
+	// the client uses a custom transport that checks and panics when the con wasn't reused.
+	_, err = client.Do(req)
+	if err == nil {
+		t.Fatal("expected to receive an error")
+	}
+}
+
+var tsCrt = []byte(`-----BEGIN CERTIFICATE-----
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:350]: will be valid for 10y

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 8]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
...
+	}
+	stderr := os.Stderr
+	cleanUp := func() {
+		os.Stderr = stderr
+		klog.LogToStderr(false)
+		writer.Close()
+	}
+	expectedOutput := "Aborting the request since ErrAbortHandler was received (net/http: abort Handler) on GET /. This might result in a EOF error on the client side"
+	defer cleanUp()
+	os.Stderr = writer
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:274]: isn't there also a buf writer without the go routine blow?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 9]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
...
+	}
+	client.Transport = tr
+
+	// act
+	_, err = client.Do(req)
+	if err == nil {
+		t.Fatal("expected to receive an error")
+	}
+
+	cleanUp()
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:326]: the defer of cleanUp does not harm?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 10]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
...
+	data, err := ioutil.ReadAll(&buf)
+	if err != nil {
+		t.Fatalf("unable to read the captured std output, err %v", err)
+	}
+	capturedOutput := string(data)
+	// maxize is the size of the expected output + 35%
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:332]: maxSize

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:332]: why 35% ?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 11]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
...
+		t.Errorf("unxpected output captured, its size = %d, exceedes = %d, output:\n%v", len(capturedOutput), maxSize, capturedOutput)
+	}
+	if !strings.Contains(capturedOutput, expectedOutput) {
+		t.Errorf("captured output doesn't contain the expected value\nexpected = %v\nactual = %v", expectedOutput, capturedOutput)
+	}
+
+	// make another req to the server
+	// the connection should be reused
+	// the client uses a custom transport that checks and panics when the con wasn't reused.
+	_, err = client.Do(req)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:344]: can one reuse a request?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go [hunk 12]
```diff
@@ -230,3 +236,163 @@ func TestErrConnKilled(t *testing.T) {
 		t.Errorf("captured output doesn't contain the expected value\nexpected = %v\nactual = %v", expectedOutput, capturedOutput)
 	}
 }
+
+type fakeTransport struct {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go:240]: panicOnNonReuseTransport

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 13]
```diff
@@ -185,6 +184,14 @@ var (
 		},
 		[]string{"verb", "group", "version", "resource", "subresource", "scope", "component", "code"},
 	)
+	// RequestAbortsTotal is a number of aborted requests with http.ErrAbortHandler
+	RequestAbortsTotal = compbasemetrics.NewCounter(
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:188]: we need to define some labels on it because aborted request won't end up in the audit logs, see `TestErrConnKilledWithAuditFilter` test.

would it be okay to have `verb, group, version and the resource` ?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:210]: yeah, would add the same we have for other request metrics.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 14]
```diff
@@ -206,6 +206,16 @@ var (
 		[]string{"filter"},
 	)
 
+	// requestAbortsTotal is a number of aborted requests with http.ErrAbortHandler
+	requestAbortsTotal = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "apiserver_request_aborts_total",
+			Help:           "Number of requests which apiserver aborted possibly due to a timeout for each group, version, verb, resource, subresource and scope",
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:213]: nit: due to a timeout<comma>, for each group ...

Otherwise it reads that it timed out for every group.

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go [hunk 15]
```diff
@@ -246,15 +244,17 @@ func (tw *baseTimeoutWriter) timeout(err *apierrors.StatusError) {
...
+		// Note from the golang's docs:
+		// If ServeHTTP panics, the server (the caller of ServeHTTP) assumes
+		// that the effect of the panic was isolated to the active request.
+		// It recovers the panic, logs a stack trace to the server error log,
+		// and either closes the network connection or sends an HTTP/2
+		// RST_STREAM, depending on the HTTP protocol. To abort a handler so
+		// the client sees an interrupted response but the server doesn't log
+		// an error, panic with the value ErrAbortHandler.
+		//
+		// We are throwing http.ErrAbortHandler deliberately so that a client is notified
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:256]: ...., but we want to supress a not helpful stacktrace in the logs.

File: staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go [hunk 16]
```diff
@@ -17,22 +17,37 @@ limitations under the License.
...
+			// an error, panic with the value ErrAbortHandler.
+			//
+			// Note that the ReallyCrash variable controls the behaviour of the HandleCrash function
+			// So it might actually crash, after calling the handlers
+			info, err := resolver.NewRequestInfo(req)
+			if err != nil {
+				metrics.RecordRequestAbort(req, nil)
+				return
+			}
+			metrics.RecordRequestAbort(req, info)
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/filters/wrap.go:50]: I'd like to see a rate limited log message about this.  We rate limit runtime.HandleError by default, so I'll add that here as a followup because @p0lyn0mial is in Europe and on vacation this week.
